iothub/device/devdoc/API Design.md

```diff
@@ -0,0 +1,255 @@
+///     "targetTemperature": 21.3,
+///     "targetHumidity": 80
+///   }
+/// },
+/// "$version" : 3
+/// 
+/// </summary>
+/// <param name="propertyActionAsTwinCollection">The action to be taken when a writeable property event is received.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public void SubscribeToWritablePropertyEvent(Action<TwinCollection> propertyActionAsTwinCollection, CancellationToken cancellationToken = default);
+```
+
+<details>
+<summary>WritableProperty</summary>
+
+```csharp
+/// <summary>
+/// Empty constructor.
+/// </summary>
+public WritablePropertyResponse() { }
+
+/// <summary>
+/// Convenience constructor for specifying the properties.
+/// </summary>
+/// <param name="propertyValue">The unserialized property value.</param>
+/// <param name="ackCode">The acknowledgement code, usually an HTTP Status Code e.g. 200, 400.</param>
+/// <param name="ackVersion">The acknowledgement version, as supplied in the property update request.</param>
+/// <param name="ackDescription">The acknowledgement description, an optional, human-readable message about the result of the property update.</param>
+public WritablePropertyResponse(object propertyValue, int ackCode, long ackVersion, string ackDescription = null)
+{
+    PropertyValue = propertyValue;
+    AckCode = ackCode;
+    AckVersion = ackVersion;
+    AckDescription = ackDescription;
+}
+
+/// <summary>
+/// The unserialized property value.
+/// </summary>
+[JsonProperty("value")]
+public dynamic PropertyValue { get; set; }
+
+/// <summary>
+/// The acknowledgement code, usually an HTTP Status Code e.g. 200, 400.
+/// </summary>
+[JsonProperty("ac")]
+public int AckCode { get; set; }
+
+/// <summary>
+/// The acknowledgement version, as supplied in the property update request.
+/// </summary>
+[JsonProperty("av")]
+public long AckVersion { get; set; }
+
+/// <summary>
+/// The acknowledgement description, an optional, human-readable message about the result of the property update.
+/// </summary>
+[JsonProperty("ad", DefaultValueHandling = DefaultValueHandling.Ignore)]
+public string AckDescription { get; set; }
+```
+</details>
+
+
+### Telemetry
+
+```csharp
+/// <summary>
+/// Send a single instance of telemetry.
+/// </summary>
+/// <param name="telemetryName">The name of the telemetry, as defined in the DTDL interface.</param>
+/// <param name="telemetryValue">The telemetry payload, in the format defined in the DTDL interface.</param>
+/// <param name="componentName">The name of the component in which the telemetry is defined. Can be null for telemetry defined under the root interface.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task SendTelemetryAsync(string telemetryName, dynamic telemetryValue, string componentName = default, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Send a batched instance of telemetry.
```

Comment:
Does MQTT support batched telemetry in C#?  At the protocol I know AMQP has 1st class support for it, but MQTT doesn't.

If C#/MQTT doesn't support this somehow and short of a critical need, I'd strongly recommend removing this from API surface.

I worry having an AMQP only construct when we're pushing MQTT as preferred protocol isn't clear.  FWIW C absolutely will not have a batched mode for PnP telemetry.

Reply:
no, not at the protocol level.

In this case, we're putting more than one property in the same message body.

Reply:
Yes, as David mentioned, this isn't batching multiple messages, but rather creating a single message that contains multiple teletry key-value pairs, put into the same message payload.
.NET doesn't really support message batching over MQTT either, we have a loop-and-send implementation that doesn't really add any value.

Reply:
avoid the term "batch" in comments/ API surface